kernel/src/locking/rwlock.rs

```diff
@@ -52,17 +145,119 @@ pub type ReadLockGuardAnyTpr<'a, T, const TPR: usize> =
+    where
+        F: FnOnce(&mut T) -> Option<&mut U>,
+    {
+        // Ensure the original guard is never dropped
+        let mut orig = ManuallyDrop::new(orig);
+        let rwlock = orig.rwlock;
+
+        // Move the original IRQ state out of drop. We need to do this
+        // before calling the closure to make the borrow checker happy.
+        // SAFETY: we are really reading from a reference, so the source is
+        // valid. The original guard will not be dropped since is is behind
+        // `ManuallyDrop`, so `drop` will only happen once. If we bail from
+        // this function, we'll pass the new copy of the guard to forget()
+        // to avoid dropping it, ensuring only the original guard invokes
+        // drop.
+        let _irq_state = unsafe { core::ptr::read(&orig._irq_state) };
+
+        let Some(value) = f(&mut *orig) else {
+            // Do not drop the new IRQ guard, since we keep the old one.
+            // Otherwise we'll drop the same guard twice.
+            core::mem::forget(_irq_state);
+            return Err(ManuallyDrop::into_inner(orig));
+        };
+
+        Ok(RawWriteLockGuard {
+            rwlock,
+            data: NonNull::from(value),
+            _variance: PhantomData,
+            _irq_state,
+        })
+    }
+}
+
+impl<'a, T, I: IrqLocking> RawWriteLockGuard<'a, T, I> {
+    pub fn map_split<U, V, F>(
+        orig: Self,
+        f: F,
+    ) -> (RawWriteLockGuard<'a, U, I>, RawWriteLockGuard<'a, V, I>)
+    where
+        F: FnOnce(&mut T) -> (&mut U, &mut V),
+    {
+        // Ensure the original guard is never dropped
+        let mut orig = ManuallyDrop::new(orig);
+
+        // Move the original IRQ state out of drop.
+        // SAFETY: we are really reading from a reference, so the source
+        // pointer is safe. The original guard is behind `ManuallyDrop`,
+        // so only the copy we just make will invoke drop.
+        let _irq_state = unsafe { core::ptr::read(&orig._irq_state) };
+
+        // Increase the writer count by 1.
+        let rwlock = orig.rwlock;
+        rwlock.fetch_add(1 << RW_BITS, Ordering::Release);
+
+        let (a, b) = f(&mut *orig);
+        (
+            RawWriteLockGuard {
+                rwlock,
+                data: NonNull::from(a),
+                _variance: PhantomData,
+                _irq_state,
+            },
+            RawWriteLockGuard {
+                rwlock,
+                data: NonNull::from(b),
+                _variance: PhantomData,
+                // Acquire a second IRQ guard
+                _irq_state: I::acquire_lock(),
+            },
+        )
+    }
+}
+
 /// Implements the behavior of the [`WriteLockGuard`] when it is dropped
 impl<T, I> Drop for RawWriteLockGuard<'_, T, I> {
     fn drop(&mut self) {
-        // There are no readers - safe to just set lock to 0
-        self.rwlock.store(0, Ordering::Release);
+        self.rwlock.fetch_sub(1 << RW_BITS, Ordering::Release);
```

Comment:
```suggestion
        self.rwlock.fetch_sub(compose_val(0, 1), Ordering::Release);
```
This makes the manipulation consistent with how the write lock is acquired.